src/QMCHamiltonians/OperatorBase.h

```diff
@@ -95,356 +97,356 @@ struct OperatorBase : public QMCTraits
+   */
+  virtual void updateSource(ParticleSet& s);
 
-  /** named values to  the property list
+  /**
+   * set the values evaluated by this object to plist
    * @param plist RecordNameProperty
    *
-   * Previously addObservables but it is renamed and a non-virtial function.
+   * Default implementation is to assign Value which is updated
+   * by evaluate  function using myIndex.
    */
-  inline void addValue(PropertySetType& plist)
-  {
-    if (!UpdateMode[COLLECTABLE])
-      myIndex = plist.add(myName.c_str());
-  }
+  // TODO() noexcept
+  virtual void setObservables(PropertySetType& plist);
 
-  /** named values to  the property list
+  /**
+   * named values to  the property list
    * @param plist RecordNameProperty
    * @param collectables Observables that are accumulated by evaluate
-   *
    * Default implementaton uses addValue(plist)
    */
-  virtual void addObservables(PropertySetType& plist, BufferType& collectables) { addValue(plist); }
+  virtual void addObservables(PropertySetType& plist, BufferType& collectables);
 
-  /*** add to observable descriptor for hdf5
+  /**
+   * add to observable descriptor for hdf5
    * @param h5desc contains a set of hdf5 descriptors for a scalar observable
    * @param gid hdf5 group to which the observables belong
    *
    * The default implementation is to register a scalar for this->Value
    */
   virtual void registerObservables(std::vector<ObservableHelper>& h5desc, hid_t gid) const;
 
-  /*** add to collectables descriptor for hdf5
+  /**
+   * add to collectables descriptor for hdf5
    * @param h5desc contains a set of hdf5 descriptors for a scalar observable
    * @param gid hdf5 group to which the observables belong
    *
    * The default implementation does nothing. The derived classes which compute
    * big data, e.g. density, should overwrite this function.
    */
-  virtual void registerCollectables(std::vector<ObservableHelper>& h5desc, hid_t gid) const {}
+  virtual void registerCollectables(std::vector<ObservableHelper>& h5desc, hid_t gid) const;
 
-  /** set the values evaluated by this object to plist
-   * @param plist RecordNameProperty
-   *
-   * Default implementation is to assign Value which is updated
-   * by evaluate  function using myIndex.
-   */
-  virtual void setObservables(PropertySetType& plist) { plist[myIndex] = Value; }
+  // TODO add documentation
+  virtual void getRequiredTraces(TraceManager& tm);
 
-  virtual void setParticlePropertyList(PropertySetType& plist, int offset) { plist[myIndex + offset] = Value; }
+  // TODO add documentation
+  virtual void setParticlePropertyList(PropertySetType& plist, int offset);
 
-  //virtual void setHistories(Walker<Return_t, ParticleSet::ParticleGradient_t>& ThisWalker)
-  virtual void setHistories(Walker_t& ThisWalker) { tWalker = &(ThisWalker); }
+  // TODO add documentation
+  virtual void setRandomGenerator(RandomGenerator_t* rng);
```

Comment:
For docs.  In the batched drivers this is called to set the rng to the crowd rng. So multiple hamiltonians that have the expectation of being called serially share an rng. If you then have nested threads or tasks over those operators you are creating a potential race on the rng.